### PR TITLE
Added sample for LiDAR-based object detection with PointPillars

### PR DESCRIPTION
**Pull request by fabianoboril**
# Description
This PR adds a new sample based for LiDAR-based object detection using PointPillars

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Command Line (as described in README)
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode

# Checklist for Moving samples:
Links and Details can be found in the samples WG Teams Files. 

- [ ] Review sample design with domain reviewers https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts 
- [ ] Implement coding guidelines and ensure code quality.
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com
- [x] Adhere to readme template 
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Enforce format via clang-format config file
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth or Joe Oster. (GitHub User: JoeOster)
- [x] Tested using Dev Cloud when applicable
- [ ] Implement fixes for ONSAM Jiras
- [x] If you have new dependencies/binaries, inform Samples Project Manager Swapna R Dontharaju (@srdontha) or @JoeOster

### CODE REVIEW

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    } else if (line.find("DATA") != std::string::npos) {
+      parse_data = true;
+    }
+  }
+
+  return number_of_points;
+}
+
+int main(int argc, char *argv[]) {
+  boost::program_options::options_description desc("Allowed options");
```
Comment (pmpeter1) [inline AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp:51]: The boost library is a new dependence I have not seen in other sample code I've reviewed.  Please check to see if we have it approved in the IP plan.

**Review by JoeOster — APPROVED** [review]
approving for ci run only

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  sycl::free(dev_y_coors_for_sub_, queue);
+}
+
+void PreProcess::DoPreProcess(const float *dev_points, const int in_num_points, int *dev_x_coors, int *dev_y_coors,
+                              float *dev_num_points_per_pillar, float *dev_pillar_x, float *dev_pillar_y,
+                              float *dev_pillar_z, float *dev_pillar_i, float *dev_x_coors_for_sub_shaped,
+                              float *dev_y_coors_for_sub_shaped, float *dev_pillar_feature_mask,
+                              int *dev_sparse_pillar_map, int *host_pillar_count) {
+  sycl::queue queue = devicemanager::GetCurrentQueue();
+  queue.memset(dev_pillar_x_in_coors_, 0, grid_y_size_ * grid_x_size_ * max_num_points_per_pillar_ * sizeof(float))
```
Comment (pmpeter1) [inline AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp:187]: Why do you .wait() after every operation   There is no dependence between then so you should be able to launch them all and then only wait at the end for all off them to complete.

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+    h.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, max_num_pillars_) * sycl::range<3>(1, 1, max_num_points_per_pillar_),
+                          sycl::range<3>(1, 1, max_num_points_per_pillar_)),
+        [=](sycl::nd_item<3> item_ct1) {
+          MakeExtraNetworkInputKernel(dev_x_coors_for_sub_ct0, dev_y_coors_for_sub_ct1, dev_num_points_per_pillar,
+                                      dev_x_coors_for_sub_shaped, dev_y_coors_for_sub_shaped, dev_pillar_feature_mask,
+                                      max_num_points_per_pillar_ct6, item_ct1);
+        });
+  });
```
Comment (pmpeter1) [inline AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp:290]: Where is the .wait() synchronization needed here?

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp (hunk 1 above)
Comment (fabianoboril) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp:51]: The use of boost was approved and is also mentioned in the 3rd-party-licenses

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp (hunk 2 above)
Comment (fabianoboril) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp:187]: Good point. Updated that, to wait only after all memset/memcpy operations. Similar updates were made in all files.

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp (hunk 3 above)
Comment (fabianoboril) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/pointpillars/preprocess.cpp:290]: In fact, we used an in_order_queue. We updated it to an unrestricted queue now, and added wait() operations for synchronization.

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp (hunk 1 above)
Comment (JoeOster) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/src/main.cpp:51]: Correct

**Review by JoeOster — APPROVED** [review]
approving for ci

**Comment by JoeOster** [comment]
Currently, Linux will fail ci due to openVINO not installed on CI machines
.
Currently waiting on Readme changes

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/README.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+![Overview](data/point_pillars_overview.png)
+
+
+By default the application will use 'host' as the Intel® oneAPI execution device and CPU for Intel® Distribution of OpenVINO™ toolkit inferencing part. The Intel® oneAPI execution device and inferencing device are displayed in the output along with elapsed time of each of the five steps described above. For more details refer to section: [Execution Options for the Sample Program](#execution-options-for-the-sample-program).
```
Comment (JoeOster) [inline AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/README.md]: his section refers to the 5 steps above... reading through it, it's hard to pull 5 steps out (I get 4). So people unfamiliar with the technology may struggle to comprehend. Move the steps into an organized list format to enhance clarity. Please note I made some edits to your GitHub version of the readme for grammar and punctuation.

**Review by JoeOster — CHANGES_REQUESTED** [review]

File: AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/README.md (hunk 4 above)
Comment (fabianoboril) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/README.md]: Hi Joe,
I updated the readme to provide this in a list form and aligned it with our illustration of the algorithm (6 steps).

Comment (JoeOster) [reply AI-and-Analytics/End-to-end-Workloads/LidarObjectDetection-PointPillars/README.md]: That looks great thank you

**Comment by JoeOster** [comment]
Per Fabian - we tested this code last Friday (03/05) in the devcloud for DG1, integrated graphics, Core and Xeon. Waiving the CI failure as its an issue of not having OpenVINO installed

**Review by JoeOster — APPROVED** [review]

**Pull request merged**